src/relay/mod.rs

```diff
@@ -63,19 +65,28 @@ where
             self.limiter,
         ));
         loop {
-            let (tcp_stream, _) = tcp_listener.accept().await?;
-            let http = http.clone();
-            tokio::task::spawn(async move {
-                if let Err(http_err) = Http::new()
-                    .http1_keep_alive(true)
-                    .serve_connection(tcp_stream, http)
-                    .with_upgrades()
-                    .await
-                {
-                    eprintln!("Error while serving HTTP connection: {}", http_err);
-                }
-            });
+            select! {
+                _ = signal::ctrl_c() => {
+                    log::info!("shutting down relay gracefully");
+                    break;
```

Comment:
no need to close the listener here?
something like
```rust
tcp_lister.close()...
```

Reply:
tokio tcp_listener doesn't implement close or shutdown function to terminate the listener.
but using this tokio `select` simultaneously awaits for both the two actions and whenever one hits first it goes with it 
So once the Ctrl+C signal is received, the listener stops accepting new TCP connections, and the application exits the loop

Reply:
> So once the Ctrl+C signal is received, the listener stops accepting new TCP connections, and the application exits the loop

Yes, but this is not graceful. All in-flight requests will be forcefully closed as well.

There are two alternatives i can think of:
1. see the example at tokio mini-redis https://github.com/tokio-rs/mini-redis/blob/e186482ca00f8d884ddcbe20417f3654d03315a4/src/server.rs#L172-L197

2. If client can handle it, then it is probably OK to not shutdown it gracefully.